File: src/monnef/core/utils/RandomHelper.java
/*
 * Jaffas and more!
 * author: monnef
 */

package monnef.core.utils;

import java.util.Random;

public class RandomHelper {
    public static Random rand = new Random();

    public static float generateRandomFromInterval(float min, float max) {
        float size = max - min;
        if (size < 0) throw new IllegalArgumentException();
        return rand.nextFloat() * size + min;
    }

    public static float[] randomPositionInCircle(float radius) {
        double angle = rand.nextDouble();
        double x = radius * Math.cos(angle);
        double y = radius * Math.sin(angle);
        return new float[]{(float) x, (float) y};
    }

    public static float generateRandomFromSymmetricInterval(float len) {
        return generateRandomFromInterval(-len, len);
    }

    public static int generateRandomFromInterval(int min, int max) {
        int size = max - min;
        if (size < 0) throw new IllegalArgumentException();
        return rand.nextInt(size + 1) + min;
    }

    public static int generateRandomFromSymmetricInterval(int len) {
        return generateRandomFromInterval(-len, len);
    }

    public static boolean rollPercentBooleanDice(int percentsOfSuccess) {
        return rand.nextFloat() < percentsOfSuccess / 100f;
    }
}

File: src/main/java/monnef/core/utils/RandomHelper.scala
/*
 * Jaffas and more!
 * author: monnef
 */
package monnef.core.utils

import java.util.Random

object RandomHelper {
  def generateRandomFromInterval(min: Float, max: Float): Float = {
    val size: Float = max - min
    if (size < 0) throw new IllegalArgumentException
    rand.nextFloat * size + min
  }

  def randomPositionInCircle(radius: Float): Array[Float] = {
    val angle: Double = rand.nextDouble
    val x = radius * Math.cos(angle)
    val y = radius * Math.sin(angle)
    Array[Float](x.asInstanceOf[Float], y.asInstanceOf[Float])
  }

  def generateRandomFromSymmetricInterval(len: Float): Float = generateRandomFromInterval(-len, len)

  def generateRandomFromInterval(min: Int, max: Int): Int = {
    val size: Int = max - min
    if (size < 0) throw new IllegalArgumentException
    rand.nextInt(size + 1) + min
  }

  def generateRandomFromSymmetricInterval(len: Int): Int = {
    generateRandomFromInterval(-len, len)
  }

  def rollPercentBooleanDice(percentsOfSuccess: Int): Boolean = {
    rand.nextFloat < percentsOfSuccess / 100f
  }

  def generateRandomFromBaseAndSpread(base: Float, spread: Float): Float = {
    base + generateRandomFromInterval(-spread, spread)
  }

  def generateRandomFromBaseAndSpread(base: Int, spread: Int): Int = {
    base + generateRandomFromInterval(-spread, spread)
  }

  var rand: Random = new Random

  def randomlySelectOne[T](data: Seq[(T, Int)]): T = {
    val fixedData = data.filter(_._2 > 0)
    val totalValue = fixedData.foldLeft(0) { case (acc, (item, value)) => acc + value}
    val randomValue = RandomHelper.rand.nextInt(totalValue)
    var currentValue = 0
    var itemIndex = 0
    while (currentValue < randomValue) {
      currentValue += fixedData(itemIndex)._2
      if (currentValue <= randomValue) itemIndex += 1
    }
    fixedData(itemIndex)._1
  }
}